File: app/src/main/java/com/yeraygarcia/recipes/ShoppingListFragment.java
package com.yeraygarcia.recipes;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;

import com.yeraygarcia.recipes.adapter.ShoppingListAdapter;
import com.yeraygarcia.recipes.database.entity.custom.UiShoppingListItem;
import com.yeraygarcia.recipes.util.Debug;
import com.yeraygarcia.recipes.util.ShortDividerItemDecoration;
import com.yeraygarcia.recipes.viewmodel.RecipeViewModel;

public class ShoppingListFragment extends Fragment {

    private FragmentActivity mParentActivity;
    private RecipeViewModel mViewModel;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mParentActivity = getActivity();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        Debug.d(this, "onCreateView(inflater, container, savedInstanceState)");
        View rootView = inflater.inflate(R.layout.fragment_shopping_list, container, false);

        setHasOptionsMenu(true);

        mViewModel = ViewModelProviders.of(this).get(RecipeViewModel.class);

        setupAddItemViews(rootView);

        ShoppingListAdapter shoppingListAdapter = new ShoppingListAdapter(mParentActivity, mViewModel);
        RecyclerView shoppingListRecyclerView = rootView.findViewById(R.id.recyclerview_shopping_list);
        shoppingListRecyclerView.setLayoutManager(new LinearLayoutManager(mParentActivity));
        shoppingListRecyclerView.addItemDecoration(new ShortDividerItemDecoration(mParentActivity, DividerItemDecoration.VERTICAL, 16));
        shoppingListRecyclerView.setAdapter(shoppingListAdapter);

        mViewModel.getShoppingListItems().observe(this, shoppingListAdapter::setShoppingListItems);
        mViewModel.getUnits().observe(this, units -> {/* TODO This is a hack :-( */});

        // recognize when a user swipes an item
        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(final RecyclerView.ViewHolder viewHolder, int swipeDir) {
                int position = viewHolder.getAdapterPosition();
                UiShoppingListItem shoppingListItem = shoppingListAdapter.getShoppingListItems().get(position);
                mViewModel.removeFromShoppingList(shoppingListItem);
            }
        }).attachToRecyclerView(shoppingListRecyclerView);


        return rootView;
    }

    private void setupAddItemViews(View rootView) {
        EditText editTextNewItem = rootView.findViewById(R.id.edittext_new_item);
        ImageButton imageButtonAddItem = rootView.findViewById(R.id.imagebutton_add_item);

        imageButtonAddItem.setOnClickListener(view -> {
            String newItem = editTextNewItem.getText().toString();
            mViewModel.addItemToShoppingList(newItem);
            editTextNewItem.setText("");
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        Debug.d(this, "onPause()");
        mViewModel.persistDraft();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        Debug.d(this, "onCreateOptionsMenu(menu, inflated)");
        inflater.inflate(R.menu.options_menu_shopping_list, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.clear_completed_button:
                mViewModel.clearCompletedFromShoppingList();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/ShoppingListFragment.kt
package com.yeraygarcia.recipes

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.*
import android.widget.ArrayAdapter
import com.yeraygarcia.recipes.adapter.ShoppingListAdapter
import com.yeraygarcia.recipes.util.ShortDivider
import com.yeraygarcia.recipes.util.SpaceTokenizer
import com.yeraygarcia.recipes.viewmodel.RecipeViewModel
import kotlinx.android.synthetic.main.fragment_shopping_list.view.*
import timber.log.Timber

class ShoppingListFragment : Fragment() {

    lateinit var viewModel: RecipeViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        Timber.d("onCreateView(inflater, container, savedInstanceState)")
        val rootView = inflater.inflate(R.layout.fragment_shopping_list, container, false)

        setHasOptionsMenu(true)

        activity?.let { activity ->

            if (!::viewModel.isInitialized) {
                Timber.d("ViewModel was empty. Initializing...")
                viewModel = ViewModelProviders.of(activity).get(RecipeViewModel::class.java)
            }

            val newItemAdapter = ArrayAdapter(
                activity,
                android.R.layout.simple_dropdown_item_1line,
                mutableListOf<String>()
            )
            val newItemEditText = rootView.editTextNewItem.apply {
                setAdapter(newItemAdapter)
                setTokenizer(SpaceTokenizer())
            }

            rootView.imageButtonAddItem.setOnClickListener {
                val newItem = newItemEditText.text.toString()
                viewModel.addItemToShoppingList(newItem)
                newItemEditText.setText("")
            }

            val shoppingListAdapter = ShoppingListAdapter(activity, viewModel)
            val shoppingListRecyclerView = rootView.recyclerViewShoppingList.apply {
                layoutManager = LinearLayoutManager(activity)
                addItemDecoration(ShortDivider(activity, DividerItemDecoration.VERTICAL, 16))
                adapter = shoppingListAdapter
            }

            viewModel.shoppingListItems.observe(activity, Observer {
                shoppingListAdapter.shoppingListItems = it ?: emptyList()
            })
            viewModel.units.observe(activity, Observer {
                /* RecipeViewModel.getUnits() needs to be observed in order to fill RecipeViewModel.mUnits
                 * which is used in RecipeViewModel.getUnitIdByName()
                 * which is used in RecipeViewModel.addItemToShoppingList
                 * which is used further up */
            })
            viewModel.unitsAndIngredientNames.observe(activity, Observer {
                newItemAdapter.clear()
                newItemAdapter.addAll(it ?: mutableListOf())
            })

            // recognize when a user swipes an item
            ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT) {
                override fun onMove(
                    recyclerView: RecyclerView,
                    viewHolder: RecyclerView.ViewHolder,
                    target: RecyclerView.ViewHolder
                ): Boolean {
                    return false
                }

                override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {
                    val position = viewHolder.adapterPosition
                    val shoppingListItem = shoppingListAdapter.shoppingListItems[position]
                    viewModel.removeFromShoppingList(shoppingListItem)
                }
            }).attachToRecyclerView(shoppingListRecyclerView)
        }

        return rootView
    }

    override fun onPause() {
        super.onPause()
        Timber.d("onPause()")
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        Timber.d("onCreateOptionsMenu(menu, inflated)")
        inflater?.inflate(R.menu.options_menu_shopping_list, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return when (item?.itemId) {
            R.id.clear_completed_button -> {
                viewModel.clearCompletedFromShoppingList()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
